_posts/2021-09-26-where-the-store-is-stored.md

```diff
@@ -0,0 +1,292 @@
+security issue.
+
+For library or framework locations such as our React `useState`, the big
+advantage is probably the strong integration with the corresponding library.
+Even bigger store solutions like Redux come with their binding like
+`react-redux` that does the job for us. No need to handle the Context / Provider
+stuff I mentioned. No issue with server-side rendering: the store being "in
+React runtime", it will be scoped to the `ReactDOM.render` or
+`ReactDOMServer.renderToString` (which will be called once for every http
+requests).
+
+I'm beginning to think this deep integration with view librairies in a way that
+makes server and browser contexts similar is a big factor in the success of
+Redux and others (Mobx, etc.).
+
+The main issue is however that we're locked to the corresponding library /
+framework.  Let's say you have an SPA where you want React to really only be
+about the views, and that you need to read / update / subscribe to store change
+in other contexts.  You just can't, it's React or bust. Back at square one.
+
+Let's give another chance to our programmatic store sharing solution, ie.
+"crafting a store as soon as the app starts and then pass it along to every
+subsequent bit of code that needs it through a sort of a handmade dependency
+injection".
+
+It's clearly not very convenient to have to pass our store programaticaly to
+every part of the codebase as it creates a lot of boilerplate, but at least
+we face no issue with SSR if we craft one store per request server-side, and one
+store when the app starts browser-side. It is very flexible as well: we
+can do whatever we want with our store and use it everywhere in the codebase.
+
+I didn't mention testing as all three locations look kind of similar to me in
+terms of how easy it is to test code that rely on them. We can use full module
+mocking (eg. with `jest.mock`). This is especially useful for option 1. We can
+also use "programmatic mocks", more or less handcrafted (option 2 and option 3)
+by crafting fake stores and passing them to the code that use them (through
+injection or by wrapping a fake React `Provider`).
+
+# No one location to rule them all
+
+There is no perfect solution to host stores with Javascript.  It's all about
+compromises. The module location is very convenient as long as you don't have to
+render your app server-side. We don't have threads or or the luxury of a
+single runtime per user there. You may find solutions exploring low level [NodeJS
+primitives](https://nodejs.org/api/worker_threads.html) or [V8
+functionalities](https://github.com/laverdet/isolated-vm) deep features to
+isolate runtimes, but it's certainly not out of the box.
+
+The library / framework location is very convenient but leaves no
+room for code outside of the library / framework pattern which is a big issue in my
+opinion for big and/or old codebases that have to support multiple technologies,
+historicities and teams.
```

Comment:
historicities ? :)

Reply:
yep yep https://en.wikipedia.org/wiki/Historicity

Reply:
`historicités` in french